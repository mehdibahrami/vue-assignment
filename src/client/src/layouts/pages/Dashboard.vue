<template>
  <main class="main">
    <div class="container">
      <!-- Column grid -->
      <div class="grid-cols-1 grid-lg-cols-2">
        <!-- Left column -->
        <div class="grid-cols-1">
          <Card :hasPadding="false">
            <LeafletMap
              :lat="data.gps.lat"
              :lng="data.gps.lng"
              :soc="data.soc"
            />
          </Card>
        </div>
        <!-- Right column -->
        <div class="grid-cols-1">
          <!-- Column grid -->
          <div class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2">
            <!-- Left column - Speed gauge chart -->
            <Card title="Speed">
              <GaugeChart
                :data="data.speed"
                :options="speedGaugeChartOptions"
              />
            </Card>
            <!-- Right column - SoC -->
            <Card title="State of Charge">
              <GaugeChart :data="data.soc" :options="socGaugeChartOptions" />
            </Card>
          </div>
          <!-- Column grid -->
          <div class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2">
            <!-- Left column - Energy -->
            <StatisticsCard title="Energy" :statistic="data.energy" unit="kW" />
            <!-- Right column - Odometer -->
            <StatisticsCard title="Odometer" :statistic="data.odo" unit="km" />
          </div>
        </div>
      </div>
      <!-- Column grid -->
      <div class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2 grid-top-margin">
        <!-- Left column - Speed line chart -->
        <Card title="Speed (km/h)">
          <LineChart
            title="Speed"
            :data="{ x: data.time, y: data.speed }"
            :options="speedLineChartOptions"
            :interval="speedLineChartInterval"
          />
        </Card>
      </div>
    </div>
  </main>
</template>

<script>
import {
  Card,
  LeafletMap,
  GaugeChart,
  StatisticsCard,
  LineChart,
} from "@/components";
import { SPEED_LINE_CHART_INTERVAL } from "@/constants";
import {
  speedGaugeChartOptions,
  socGaugeChartOptions,
  speedLineChartOptions,
} from "@/components/charts/chartsOptions.js";

export default {
  name: "Dashboard",
  components: {
    Card,
    LeafletMap,
    GaugeChart,
    StatisticsCard,
    LineChart,
  },
  props: {
    data: Object,
  },
  data() {
    return {
      speedGaugeChartOptions,
      socGaugeChartOptions,
      speedLineChartOptions,
      speedLineChartInterval: SPEED_LINE_CHART_INTERVAL,
    };
  },
};
</script>

<style scoped lang="scss">
.main {
  margin-top: -6rem;
}

.container {
  max-width: 80rem;
  margin-left: auto;
  margin-right: auto;
  padding-bottom: 2rem;
  padding-left: 1rem;
  padding-right: 1rem;
}

@media (min-width: 640px) {
  .container {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }
}

@media (min-width: 1024px) {
  .container {
    padding-left: 2rem;
    padding-right: 2rem;
  }
}
</style>
