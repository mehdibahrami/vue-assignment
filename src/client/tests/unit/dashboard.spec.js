import { shallowMount } from "@vue/test-utils";
import Dashboard from "@/layouts/pages/Dashboard.vue";

describe("Dashboard", () => {
  let wrapper;

  it("renders correctly", () => {
    // shallow mounting the main component
    // Stub card component replacing with its template to load slotted components
    wrapper = shallowMount(Dashboard, {
      global: {
        stubs: {
          Card: {
            template: '<div class="card-body"><slot></slot></div>',
          },
        },
      },
      props: {
        data: {
          energy: 53.2,
          gps: {
            lat: 52.093448638916016,
            lng: 5.117378234863281,
          },
          odo: 88526.414,
          soc: 72.8,
          speed: 4,
          time: 1511436339000,
        },
      },
    });

    expect(wrapper.findComponent({ name: "LeafletMap" }).exists()).toBeTruthy();
    expect(wrapper.findComponent({ name: "GaugeChart" }).exists()).toBeTruthy();
    expect(
      wrapper.findComponent({ name: "StatisticsCard" }).exists()
    ).toBeTruthy();
    expect(wrapper.findComponent({ name: "LineChart" }).exists()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });
});
