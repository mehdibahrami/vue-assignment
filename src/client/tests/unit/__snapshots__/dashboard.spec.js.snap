// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correctly 1`] = `
<main
  class="main"
>
  <div
    class="container"
  >
    <!-- Column grid -->
    <div
      class="grid-cols-1 grid-lg-cols-2"
    >
      <!-- Left column - Map -->
      <div
        class="grid-cols-1"
      >
        <div
          class="card-body"
          haspadding="false"
        >
          
          <leaflet-map-stub
            lat="52.093448638916016"
            lng="5.117378234863281"
            soc="72.8"
          />
          
        </div>
      </div>
      <!-- Right column -->
      <div
        class="grid-cols-1"
      >
        <!-- Column grid -->
        <div
          class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2"
        >
          <!-- Left column - Speed gauge chart -->
          <div
            class="card-body"
            title="Speed"
          >
            
            <gauge-chart-stub
              data="4"
              options="[object Object]"
            />
            
          </div>
          <!-- Right column - SoC gauge chart -->
          <div
            class="card-body"
            title="State of Charge"
          >
            
            <gauge-chart-stub
              data="72.8"
              options="[object Object]"
            />
            
          </div>
        </div>
        <!-- Column grid -->
        <div
          class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2"
        >
          <!-- Left column - Energy -->
          <statistics-card-stub
            statistic="53.2"
            title="Energy"
            unit="kW"
          />
          <!-- Right column - Odometer -->
          <statistics-card-stub
            statistic="88526.414"
            title="Odometer"
            unit="km"
          />
        </div>
      </div>
    </div>
    <!-- Column grid -->
    <div
      class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2 grid-top-margin"
    >
      <!-- Left column - Speed line chart -->
      <div
        class="card-body"
        title="Speed (km/h)"
      >
        
        <line-chart-stub
          data="[object Object]"
          interval="100"
          options="[object Object]"
          setcolor="function setSpeedLineChartColor() {
  if (this.data.y) {
    if (this.data.y <= _constants.SPEED_LIMIT) {
      this.$refs.chart.updateOptions({
        colors: [_export.default.green]
      });
    } else {
      this.$refs.chart.updateOptions({
        colors: [_export.default.red]
      });
    }
  }
}"
          title="Speed"
        />
        
      </div>
      <!-- Right column - SoC line chart -->
      <div
        class="card-body"
        title="State of Charge (%)"
      >
        
        <line-chart-stub
          data="[object Object]"
          interval="100"
          options="[object Object]"
          setcolor="function setSocLineChartColor() {
  if (this.data.y) {
    if (this.data.y >= _constants.MID_CHARGE) {
      this.$refs.chart.updateOptions({
        colors: [_export.default.green]
      });
    } else if (this.data.y < _constants.MID_CHARGE && this.data.y >= _constants.LOW_CHARGE) {
      this.$refs.chart.updateOptions({
        colors: [_export.default.orange]
      });
    } else {
      this.$refs.chart.updateOptions({
        colors: [_export.default.red]
      });
    }
  }
}"
          title="SoC"
          type="area"
        />
        
      </div>
    </div>
    <!-- Column grid -->
    <div
      class="grid-cols-1 grid-lg-cols-2 grid-sm-cols-2 grid-top-margin"
    >
      <!-- Left column - Speed Violations Card -->
      <violations-card-stub
        data="[object Object]"
        title="Last Three Speed Violations"
      />
      <!-- Right column - Restricted Areas Card -->
      <violations-card-stub
        data="[object Object]"
        title="Last Three Entry into Restricted Areas"
      />
    </div>
  </div>
</main>
`;
